File: Dynamic Programming/Coin-Change.java
class Solution {
    /*
    Understand
    we can use a dp array to store the minimum num of coins are some amount i
    dp[0] = min num of coins for amount 0, we always know dp[0] is 0
    dp[n] = min num of coins for amount n
    
    for each amount i, we want to get the minimum between dp[i] and dp[i - coin]+1
    dp[i]'s answer takes into consideration previous answers (dp[i] could have its answer come from dp[i-1] or dp[i-2])
    
    Match
    Backtracking
    
    Plan
    if the amount == 0
        return 0
    create a dp array of size amount+1 to store all subproblem solutions 
    fill the dp array with the value amount+1
    set dp[0] = 0
    for i = 1 to last index in dp
        for coin in coins
            if coin > 1
                continue
            dp[i] = min(dp[i], dp[i - coin] + 1)
    if dp[amount] == amount+1 return -1
    else return dp[amount]
    
    Evaluate
    Time: O(N*amount)
    Space: O(amount)
    */
    public int coinChange(int[] coins, int amount) {
        if(amount == 0) {
            return 0;
        }
        
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        dp[0] = 0;
        for(int i = 1; i < dp.length; ++i) {
            for(int coin : coins) {
                if(coin > i) {
                    continue;
                }
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }
        return dp[amount] == amount + 1 ? -1 : dp[amount];
    }
}

File: Dynamic Programming/Coin-Change.cs
public class Solution
{
    public int CoinChange(int[] coins, int amount)
    {
        Dictionary<int, int> dp = new Dictionary<int, int>();
        foreach (int coin in coins)
        {
            dp[coin] = 1;
        }

        return helper(coins, amount, dp);
    }

    private int helper(int[] coins, int amount, Dictionary<int, int> dp)
    {
        if (amount == 0)
        {
            return 0;
        }
        else if (amount < 0)
        {
            return -1;
        }
        else if (dp.ContainsKey(amount))
        {
            return dp[amount];
        }

        int result = int.MaxValue;
        foreach (int coin in coins)
        {
            int coinAmount = helper(coins, amount - coin, dp);
            if (coinAmount == -1)
            {
                continue;
            }
            result = Math.Min(result, coinAmount + 1);
        }
        dp[amount] = (result == int.MaxValue) ? -1 : result;
        return dp[amount];
    }
}

/*

                                                coinChange(6,0) 
coins = [2], amount = 3
5+5+2 
1+1+1+1+1+1+1+1+1+1+1
2+2+2+2+2+1

                                    coinChange(6) 
        1st level: coinchange(5)    coinchange(4)   coinchange(1)
        2nd level:       1                                1      
        3rd level:     coinchange(3)coinchange(2)coinchange(-1)
                                         1           -1
4th level:coinchange(2)coinchange(1)coinchange(-2)
               1             1            -1

int minCoinAmount
int coinAmount = dp(amount - coin, numCoins)
minCoinAmount = Min(minCoinAmount, coinAmount)

dp[i] = Min(coinCount, 1+dp[i-coin]) 
*/